[GPU Process][Filters] Make SVGFilter own a vector of unique FilterEffects

shallawa opened the pull request:
#### 56cc7591e5e3cf992682119fa17f74cb3733637e
<pre>
[GPU Process][Filters] Make SVGFilter own a vector of unique FilterEffects
<a href="https://bugs.webkit.org/show_bug.cgi?id=256333">https://bugs.webkit.org/show_bug.cgi?id=256333</a>
rdar://109060794

Reviewed by Simon Fraser.

Currently, SVGFilterExpressionTerm holds a Ref&lt;FilterEffect&gt; which may be referenced
more than once by multiple SVGFilterExpressionTerms.

We need make SVGFilter own a Vector of unique FilterEffects. SVGFilterExpressionTerm
will hold an index to this Vector. This will allow building an SVGFilterTransaction
which is a set of altered unique FilterEffects identified by the index in the vector
of unique FilterEffects of SVGFilter.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::SVGFilter::create):
(WebCore::SVGFilter::SVGFilter):
(WebCore::SVGFilter::buildExpression):
(WebCore::SVGFilter::supportedFilterRenderingModes const):
(WebCore::SVGFilter::effectsOfType const):
(WebCore::SVGFilter::apply):
(WebCore::SVGFilter::createFilterStyles const):
(WebCore::SVGFilter::externalRepresentation const):
* Source/WebCore/svg/graphics/filters/SVGFilter.h:
* Source/WebCore/svg/graphics/filters/SVGFilterExpression.h:
* Source/WebCore/svg/graphics/filters/SVGFilterExpressionReference.h: Removed.
* Source/WebCore/svg/graphics/filters/SVGFilterGraph.h:
(WebCore::SVGFilterGraph::SVGFilterGraph):
(WebCore::SVGFilterGraph::nodes const):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;SVGFilter&gt;::encode):
(IPC::ArgumentCoder&lt;SVGFilter&gt;::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/263886@main">https://commits.webkit.org/263886@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/788ab41614451310a276c149e0f5ba1bd5042b43

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5889 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6061 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6248 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7440 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6255 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5886 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6280 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6022 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/8541 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5997 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6034 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5331 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7498 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3511 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5307 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13249 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5376 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5385 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/7613 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5835 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4782 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5268 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1425 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9392 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5629 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/svg/graphics/filters/SVGFilter.h:
```diff
@@ -75,6 +77,7 @@ class SVGFilter final : public Filter {
     SVGUnitTypes::SVGUnitType m_primitiveUnits;
 
     SVGFilterExpression m_expression;
+    FilterEffectVector m_effects;
```
Does the order in this vector mean anything, or is it just a way to reference the individual effects?

shallawa replied:
The order of the effects in this vector is not really important. But the order of m_expression since it is the order of applying the filter. Each SVGFilterExpressionTerm now has an index to this vector. So when encoding the SVGFilter we have to encode both m_expression and m_effects in their orders.

shallawa committed:
[GPU Process][Filters] Make SVGFilter own a vector of unique FilterEffects
https://bugs.webkit.org/show_bug.cgi?id=256333
rdar://109060794

Reviewed by Simon Fraser.

Currently, SVGFilterExpressionTerm holds a Ref<FilterEffect> which may be referenced
more than once by multiple SVGFilterExpressionTerms.

We need make SVGFilter own a Vector of unique FilterEffects. SVGFilterExpressionTerm
will hold an index to this Vector. This will allow building an SVGFilterTransaction
which is a set of altered unique FilterEffects identified by the index in the vector
of unique FilterEffects of SVGFilter.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::SVGFilter::create):
(WebCore::SVGFilter::SVGFilter):
(WebCore::SVGFilter::buildExpression):
(WebCore::SVGFilter::supportedFilterRenderingModes const):
(WebCore::SVGFilter::effectsOfType const):
(WebCore::SVGFilter::apply):
(WebCore::SVGFilter::createFilterStyles const):
(WebCore::SVGFilter::externalRepresentation const):
* Source/WebCore/svg/graphics/filters/SVGFilter.h:
* Source/WebCore/svg/graphics/filters/SVGFilterExpression.h:
* Source/WebCore/svg/graphics/filters/SVGFilterExpressionReference.h: Removed.
* Source/WebCore/svg/graphics/filters/SVGFilterGraph.h:
(WebCore::SVGFilterGraph::SVGFilterGraph):
(WebCore::SVGFilterGraph::nodes const):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<SVGFilter>::encode):
(IPC::ArgumentCoder<SVGFilter>::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/263886@main
